File: core/src/validation.ts
/**
 * Zod schemas for everything that crosses a trust boundary: imported vault
 * files, decrypted payloads, and extension messages.
 *
 * Decrypted payloads are validated too. GCM already proves the bytes were
 * written by someone holding the key, but a vault written by a future (or
 * buggy) build could still be structurally wrong, and we would rather fail
 * loudly at the boundary than propagate `undefined` into the UI.
 *
 * STRICT WHERE IT PROVES SOMETHING, LOOSE WHERE IT ONLY HURTS:
 *
 * The envelope schemas stay `.strict()`. That part is unencrypted, so an
 * unexpected key there is a real signal — corruption, or someone appending to a
 * file they cannot decrypt.
 *
 * The payload schemas are `.loose()`: known fields are validated exactly as
 * before, and unrecognised ones are carried through instead of rejected. The
 * distinction that matters is between a *malformed* payload and a *well-formed
 * one carrying a field this build has not learned about yet*. Only a Keyhole
 * holding the key can produce the latter, and rejecting it does not protect the
 * user — it locks them out of their own passwords on whichever device they
 * updated last. Keyhole now runs on four surfaces against one synced vault
 * (desktop, extension, iOS, and any older build still installed), so "newer
 * vault meets older reader" is routine rather than exotic.
 *
 * Carrying unknown fields through also means an older build cannot silently
 * strip them: it re-encrypts what it read, so a round trip through a stale
 * device does not destroy data written by a current one.
 */

import { z } from 'zod';
import {
  MAX_ATTACHMENT_BYTES,
  MAX_ATTACHMENTS_VAULT_BYTES,
  VAULT_FORMAT_ID,
} from './types.ts';
import { MAX_LENGTH, MIN_LENGTH } from './password-gen.ts';
import { VaultFormatError } from './errors.ts';

const isoDate = z.string().refine((s) => !Number.isNaN(Date.parse(s)), 'Expected an ISO-8601 timestamp');
const b64 = z.string().regex(/^[A-Za-z0-9+/]*={0,2}$/, 'Expected base64');

export const encryptedBlobSchema = z
  .object({
    ivB64: b64.min(1),
    ctB64: b64.min(1),
  })
  .strict();

export const kdfParamsSchema = z
  .object({
    algorithm: z.literal('argon2id'),
    memoryKiB: z.int().min(1024).max(2 ** 21),
    iterations: z.int().min(1).max(64),
    parallelism: z.int().min(1).max(16),
    saltB64: b64.min(16),
    keyLength: z.literal(32),
  })
  .strict();

export const generatorOptionsSchema = z
  .object({
    length: z.int().min(MIN_LENGTH).max(MAX_LENGTH),
    lowercase: z.boolean(),
    uppercase: z.boolean(),
    digits: z.boolean(),
    symbols: z.boolean(),
    excludeAmbiguous: z.boolean(),
  })
  .loose();

export const settingsSchema = z
  .object({
    autoLockMinutes: z.number().min(0.5).max(24 * 60),
    clipboardClearSeconds: z.number().min(0).max(600),
    generator: generatorOptionsSchema,
    theme: z.enum(['light', 'dark', 'system']),
    lockOnHide: z.boolean(),
    // Absent on schema-3 vaults — off by default, never opt someone in.
    breachCheckEnabled: z.boolean().default(false),
  })
  .loose();

export const totpConfigSchema = z
  .object({
    digits: z.int().min(6).max(10),
    periodSeconds: z.int().min(1).max(3600),
    algorithm: z.enum(['SHA-1', 'SHA-256', 'SHA-512']),
  })
  .loose();

export const customFieldSchema = z
  .object({
    id: z.uuid(),
    label: z.string().max(128),
    value: z.string().max(4096),
    secret: z.boolean(),
  })
  .loose();

export const attachmentSchema = z
  .object({
    id: z.uuid(),
    name: z.string().min(1).max(512),
    mimeType: z.string().min(1).max(128),
    sizeBytes: z.int().min(0).max(MAX_ATTACHMENT_BYTES),
    dataB64: b64,
  })
  .loose();

export const entrySchema = z
  .object({
    id: z.uuid(),
    // Missing on vaults written before kinds existed; treat as login.
    kind: z.enum(['login', 'note']).default('login'),
    title: z.string().max(512),
    username: z.string().max(512),
    password: z.string().max(4096),
    urls: z.array(z.string().max(2048)).max(64),
    notes: z.string().max(64 * 1024),
    tags: z.array(z.string().max(64)).max(64),
    /*
     * Nullable AND defaulted, because "absent" and "null" both occur in the wild.
     * Swift synthesises `encodeIfPresent` for optional properties, so the iOS build
     * omits these keys instead of writing an explicit null — and requiring them
     * present made every vault saved by that app unreadable here, since almost no
     * entry has a TOTP secret. iOS now writes explicit nulls (see Types.swift), but
     * vaults it already wrote have to keep opening.
     */
    folderId: z.uuid().nullable().default(null),
    totpSecret: z.string().max(512).nullable().default(null),
    /*
     * Schema-4 fields. Defaulted so a schema-3 vault opens without a rewrite:
     * null / [] match "never set", which is exactly what older entries mean.
     */
    totpConfig: totpConfigSchema.nullable().default(null),
    customFields: z.array(customFieldSchema).max(64).default([]),
    attachments: z.array(attachmentSchema).max(32).default([]),
    createdAt: isoDate,
    updatedAt: isoDate,
    passwordUpdatedAt: isoDate,
    /*
     * Defaulted, like `kind` before them: a schema-2 vault has neither field, and
     * defaulting here is what lets an older vault open without a rewrite pass.
     * The cap is enforced when writing (see `rememberPassword`), not here, so a
     * vault that somehow exceeds it still opens rather than becoming unreadable.
     */
    history: z
      .array(z.object({ password: z.string().max(4096), changedAt: isoDate }).loose())
      .max(1000)
      .default([]),
    deletedAt: isoDate.nullable().default(null),
  })
  .loose();

export const folderSchema = z
  .object({
    id: z.uuid(),
    name: z.string().min(1).max(128),
    createdAt: isoDate,
  })
  .loose();

export const tombstoneSchema = z
  .object({
    id: z.uuid(),
    kind: z.enum(['entry', 'folder']),
    deletedAt: isoDate,
  })
  .loose();

export const vaultDataSchema = z
  .object({
    // Deliberately permissive: a newer-than-supported version must reach the
    // explicit check in `migrate()` so the user gets "update Keyhole" rather
    // than a generic schema error. The value is authenticated by GCM regardless.
    schemaVersion: z.int().min(1).max(1_000_000),
    entries: z.array(entrySchema).max(100_000),
    folders: z.array(folderSchema).max(10_000),
    /*
     * Optional because this parser runs BEFORE `migrate()`: a schemaVersion-1
     * vault has no tombstones yet, and `migrate()` back-fills `[]`.
     */
    tombstones: z.array(tombstoneSchema).max(100_000).default([]),
    settings: settingsSchema,
    updatedAt: isoDate,
  })
  .loose()
  .superRefine((data, ctx) => {
    let total = 0;
    for (const entry of data.entries) {
      for (const att of entry.attachments) {
        total += att.sizeBytes;
        if (att.sizeBytes > MAX_ATTACHMENT_BYTES) {
          ctx.addIssue({
            code: 'custom',
            message: `Attachment "${att.name}" exceeds the ${MAX_ATTACHMENT_BYTES} byte per-file limit.`,
            path: ['entries'],
          });
        }
      }
    }
    if (total > MAX_ATTACHMENTS_VAULT_BYTES) {
      ctx.addIssue({
        code: 'custom',
        message: `Vault attachments exceed the ${MAX_ATTACHMENTS_VAULT_BYTES} byte total budget.`,
        path: ['entries'],
      });
    }
  });

export const vaultFileSchema = z
  .object({
    format: z.literal(VAULT_FORMAT_ID),
    // See the note on `schemaVersion`: `unlockVault` performs the explicit
    // version check so the error message is actionable.
    formatVersion: z.int().min(1).max(1_000_000),
    vaultId: z.uuid(),
    createdAt: isoDate,
    updatedAt: isoDate,
    kdf: kdfParamsSchema,
    wrappedKey: encryptedBlobSchema,
    payload: encryptedBlobSchema,
  })
  .strict();

/** Parse an untrusted vault envelope, converting Zod failures into our error type. */
export function parseVaultFile(input: unknown): z.infer<typeof vaultFileSchema> {
  const result = vaultFileSchema.safeParse(input);
  if (!result.success) {
    throw new VaultFormatError(`Not a valid Keyhole vault file: ${result.error.issues[0]?.message ?? 'unknown error'}`);
  }
  return result.data;
}

/** Parse a decrypted payload. Failure here means a corrupt or future-version vault. */
export function parseVaultData(input: unknown): z.infer<typeof vaultDataSchema> {
  const result = vaultDataSchema.safeParse(input);
  if (!result.success) {
    const issue = result.error.issues[0];
    throw new VaultFormatError(
      `Vault contents failed validation at "${issue?.path.join('.') || '<root>'}": ${issue?.message ?? 'unknown error'}`,
    );
  }
  return result.data;
}

File: ios/Sources/KeyholeCore/Validation.swift
import Foundation

enum Validation {
    private static let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoFormatterNoFrac: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func isIsoDate(_ s: String) -> Bool {
        isoFormatter.date(from: s) != nil || isoFormatterNoFrac.date(from: s) != nil || Date.parseLoose(s) != nil
    }

    static func isBase64(_ s: String) -> Bool {
        let pattern = #"^[A-Za-z0-9+/]*={0,2}$"#
        return s.range(of: pattern, options: .regularExpression) != nil
    }

    static func isUuid(_ s: String) -> Bool {
        UUID(uuidString: s) != nil
    }

    public static func parseVaultFile(_ input: Any) throws -> VaultFile {
        let data: Data
        if let d = input as? Data {
            data = d
        } else if let s = input as? String {
            guard let d = s.data(using: .utf8) else {
                throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: invalid UTF-8")
            }
            data = d
        } else if let dict = input as? [String: Any] {
            data = try JSONSerialization.data(withJSONObject: dict)
        } else if let file = input as? VaultFile {
            try validateVaultFile(file)
            return file
        } else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: unexpected type")
        }

        do {
            let file = try JSONDecoder().decode(VaultFile.self, from: data)
            try validateVaultFile(file)
            return file
        } catch let err as KeyholeError {
            throw err
        } catch {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: \(error.localizedDescription)")
        }
    }

    public static func parseVaultFileJSON(_ text: String) throws -> VaultFile {
        try parseVaultFile(text)
    }

    public static func parseVaultData(_ input: Any) throws -> VaultData {
        let data: Data
        if let d = input as? Data {
            data = d
        } else if let s = input as? String {
            guard let d = s.data(using: .utf8) else {
                throw KeyholeError.vaultFormat("Vault contents failed validation")
            }
            data = d
        } else if let dict = input as? [String: Any] {
            data = try JSONSerialization.data(withJSONObject: dict)
        } else if let vd = input as? VaultData {
            return vd
        } else {
            throw KeyholeError.vaultFormat("Vault contents failed validation at \"<root>\"")
        }

        do {
            var decoded = try JSONDecoder().decode(VaultData.self, from: data)
            try validateVaultData(&decoded)
            return decoded
        } catch let err as KeyholeError {
            throw err
        } catch {
            throw KeyholeError.vaultFormat("Vault contents failed validation: \(error.localizedDescription)")
        }
    }

    private static func validateVaultFile(_ file: VaultFile) throws {
        guard file.format == VAULT_FORMAT_ID else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad format id")
        }
        guard file.formatVersion >= 1 else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad formatVersion")
        }
        guard isUuid(file.vaultId) else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad vaultId")
        }
        guard isIsoDate(file.createdAt), isIsoDate(file.updatedAt) else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad timestamps")
        }
        try validateKdf(file.kdf)
        try validateBlob(file.wrappedKey)
        try validateBlob(file.payload)
        try assertEnvelopeShape(file)
    }

    /// Cross-field envelope rules, mirroring `assertEnvelopeShape` in the TypeScript
    /// core.
    ///
    /// Runs on parse so every reader gets them: a half-written Recovery Kit must be
    /// caught here rather than surfacing later as an unexplained decryption failure
    /// during recovery, which is the worst possible moment to discover it.
    public static func assertEnvelopeShape(_ file: VaultFile) throws {
        let hasKdf = file.recoveryKdf != nil
        let hasBlob = file.recoveryWrappedKey != nil

        guard hasKdf == hasBlob else {
            throw KeyholeError.vaultFormat(
                "Vault has an incomplete Recovery Kit: recoveryKdf and recoveryWrappedKey must both be present or both absent."
            )
        }
        if file.formatVersion < 2 && hasKdf {
            throw KeyholeError.vaultFormat("A format-\(file.formatVersion) vault cannot carry a Recovery Kit.")
        }
        if let recoveryKdf = file.recoveryKdf {
            try validateKdf(recoveryKdf)
        }
        if let recoveryWrappedKey = file.recoveryWrappedKey {
            try validateBlob(recoveryWrappedKey)
        }
    }

    private static func validateKdf(_ kdf: KdfParams) throws {
        guard kdf.algorithm == "argon2id" else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad KDF algorithm")
        }
        guard kdf.memoryKiB >= 1024, kdf.memoryKiB <= (1 << 21) else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad memoryKiB")
        }
        guard kdf.iterations >= 1, kdf.iterations <= 64 else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad iterations")
        }
        guard kdf.parallelism >= 1, kdf.parallelism <= 16 else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad parallelism")
        }
        guard isBase64(kdf.saltB64), kdf.saltB64.count >= 16 else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad salt")
        }
        guard kdf.keyLength == 32 else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad keyLength")
        }
    }

    private static func validateBlob(_ blob: EncryptedBlob) throws {
        guard isBase64(blob.ivB64), !blob.ivB64.isEmpty,
              isBase64(blob.ctB64), !blob.ctB64.isEmpty
        else {
            throw KeyholeError.vaultFormat("Not a valid Keyhole vault file: bad encrypted blob")
        }
    }

    private static func validateVaultData(_ data: inout VaultData) throws {
        guard data.schemaVersion >= 1 else {
            throw KeyholeError.vaultFormat("Vault contents failed validation at \"schemaVersion\"")
        }
        guard isIsoDate(data.updatedAt) else {
            throw KeyholeError.vaultFormat("Vault contents failed validation at \"updatedAt\"")
        }
        for entry in data.entries {
            guard isUuid(entry.id) else {
                throw KeyholeError.vaultFormat("Vault contents failed validation at \"entries.id\"")
            }
        }
    }
}

private extension Date {
    static func parseLoose(_ s: String) -> Date? {
        // Accept JS-style ISO strings that Date.parse would accept.
        let f1 = ISO8601DateFormatter()
        f1.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f1.date(from: s) { return d }
        let f2 = ISO8601DateFormatter()
        f2.formatOptions = [.withInternetDateTime]
        return f2.date(from: s)
    }
}

/// Public wrappers matching `@keyhole/core` names.
public func parseVaultFile(_ input: Any) throws -> VaultFile {
    try Validation.parseVaultFile(input)
}

public func parseVaultData(_ input: Any) throws -> VaultData {
    try Validation.parseVaultData(input)
}
